hypervisor/arch/x86/sgx.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2019 Intel Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <types.h>
+#include <errno.h>
+#include <cpufeatures.h>
+#include <cpu_caps.h>
+#include <sgx.h>
+#include <cpuid.h>
+#include <vm.h>
+#include <logmsg.h>
+
+#define SGX_OPTED_IN (MSR_IA32_FEATURE_CONTROL_SGX_GE | MSR_IA32_FEATURE_CONTROL_LOCK)
+
+/* For the static variables, which are not explicitly initialzed will be inited to 0 */
+static int32_t init_sgx_ret = 0;
+static struct epc_section pepc_sections[MAX_EPC_SECTIONS];			/* physcial epc sections */
+static struct epc_map vm_epc_maps[MAX_EPC_SECTIONS][CONFIG_MAX_VM_NUM];		/* epc resource mapping for VMs */
+
+/* Partition physcial EPC resource according to VM configurations.
+ * Build the mappings between HPA and GPA for EPT mapping later.
+ * EPC resource partition and mapping relationship will stay the same after sgx init.
+ */
+static int32_t partition_epc(void)
+{
+	uint32_t eax, ebx, ecx, edx, type;
+	uint32_t psec_id, nvm_id, mid;
+	uint64_t psec_addr = 0UL, psec_size = 0UL;
+	uint64_t vm_request_size = 0UL, free_size = 0UL, alloc_size;
+	struct acrn_vm_config *vm_config;
+	int32_t ret = 0;
+
+	for (psec_id = 0U, nvm_id = 0U; psec_id < MAX_EPC_SECTIONS && nvm_id < CONFIG_MAX_VM_NUM; ) {
+		if (vm_request_size ==0U) {
+			mid = 0U;
+			vm_config = get_vm_config(nvm_id);
+			if((((vm_config->epc.size | vm_config->epc.base) & ~PAGE_MASK) != 0UL) ||
+			   ((vm_config->load_order==SOS_VM) && (vm_config->epc.size != 0UL))) {
+				ret = -EINVAL;
+				break;
+			}
+			vm_request_size = vm_config->epc.size;
+			nvm_id++;
+		}
+		if ((free_size ==0U) && (vm_request_size != 0UL)) {
+			cpuid_subleaf(CPUID_SGX_LEAF, psec_id + CPUID_SGX_EPC_SUBLEAF_BASE, &eax, &ebx, &ecx, &edx);
+			type = eax & CPUID_SGX_EPC_TYPE_MASK;
+			if (type == CPUID_SGX_EPC_TYPE_VALID) {
+				psec_addr = (((uint64_t)ebx & CPUID_SGX_EPC_HIGH_MASK) << 32U) |
+					((uint64_t)eax & CPUID_SGX_EPC_LOW_MASK);
+				psec_size =  (((uint64_t)edx & CPUID_SGX_EPC_HIGH_MASK) << 32U) |
+					((uint64_t)ecx & CPUID_SGX_EPC_LOW_MASK);
+				if (psec_size != 0UL) {
+					pepc_sections[psec_id].base = psec_addr;
+					pepc_sections[psec_id].size = psec_size;
+					free_size = psec_size;
+				} else {
+					ret = -EINVAL;
+					break;
+				}
+			} else if (type == CPUID_SGX_EPC_TYPE_INVALID) {
+				break;
+			} else {
+				pr_err("%s: unsupport EPC type %u", __func__, type);
+				ret = -EINVAL;
+				break;
+			}
+			psec_id++;
+		}
+		if (vm_request_size != 0UL) {
+			if (vm_request_size <= free_size)
+				alloc_size = vm_request_size;
+			else
+				alloc_size = free_size;
+			vm_epc_maps[mid][nvm_id - 1U].size = alloc_size;
+			vm_epc_maps[mid][nvm_id - 1U].hpa = psec_addr + psec_size - free_size;
+			vm_epc_maps[mid][nvm_id - 1U].gpa = vm_config->epc.base + vm_config->epc.size - vm_request_size;
```

Comment:
can we do nvm_id ++ later?
using nvm_id - 1  is not perfect.

Reply:
OK